File: FrontEnd/Models/Advertisement/Advertisement.py
import json
from datetime import datetime
from typing import Union

from Models.ApiModel import ApiModel

# create, update
class AdvertisementNew(ApiModel):
    def __init__(self):
        self.sponsorId: Union[int, None]
        self.name: Union[str, None]
        self.text: Union[str, None]
        self.targetAudience: Union[str, None]
        self.media: Union[str, None]
        self.priority: Union[str, None]
        self.mediaType: Union[str, None]


class AdvertisementUpdate(ApiModel):
    def __init__(self, id: int, sponsorId: int, name: str, text: str, targetAudience: str, media: str, priority: str, mediaType: str):
        self.id = id
        self.sponsorId = sponsorId
        self.name = name
        self.text = text
        self.targetAudience = targetAudience
        self.media = media
        self.priority = priority
        self.mediaType = mediaType


class StatisticsGet(ApiModel):
    def __init__(self):
        self.from_: datetime = datetime.now()
        self.to: datetime = datetime.now()

    def to_json(self):
        dict = {
            "from": self.from_.strftime("%Y-%m-%d"),
            "to": self.to.strftime("%Y-%m-%d")
        }

        return json.dumps(dict)


class AdvertisementItem:
    def __init__(self, advertisement_dict):
        self.id = advertisement_dict["id"]
        self.name = advertisement_dict["name"]


class Advertisement:
    def __init__(self, advertisement_dict):
        self.id: int = advertisement_dict["id"]
        self.sponsorId: int = advertisement_dict["userId"]
        self.text: str = advertisement_dict["text"]
        self.name: str = advertisement_dict["name"]
        self.targetAudience: Union[str, None] = advertisement_dict["targetAudience"]
        self.media: Union[str, None] = advertisement_dict["media"]
        self.show: bool = advertisement_dict["show"]
        self.updated: bool = advertisement_dict["updated"]
        self.priority: str = advertisement_dict["priority"]
        self.mediaType: str = advertisement_dict["mediaType"]


class StatisticsEconomy:
    def __init__(self, stats_dict):
        self.id: int = stats_dict["id"]
        self.advertisementId: int = stats_dict["advertisementId"]
        self.payback: float = stats_dict["payback"]
        self.pricePerClick: float = stats_dict["pricePerClick"]
        self.totalPrice: float = stats_dict["totalPrice"]
        self.income: float = stats_dict["income"]
        self.created: str = stats_dict["created"]

    @staticmethod
    def unpack(stats) -> list[any] | None:
        return [StatisticsEconomy(stat) for stat in stats]

class StatisticsEngagement:
    def __init__(self, stats_dict):
        self.id: int = stats_dict["id"]
        self.advertisementId: int = stats_dict["advertisementId"]
        self.viewCount: int = stats_dict["viewCount"]
        self.averageStayInSeconds: int = stats_dict["averageStayInSeconds"]
        self.clickCount: int = stats_dict["linkClickCount"]
        self.peoplePercentage: int = stats_dict["peoplePercentage"]
        self.created: str = stats_dict["created"]

    @staticmethod
    def unpack(stats) -> list[any] | None:
        return [StatisticsEngagement(stat) for stat in stats]

File: WebApi/WebApi.Models/Models/Sponsor/Advertisement.cs
#nullable enable
using System.Text.Json.Serialization;
using WebApi.Enums.Enums.Advertisement;
using WebApi.Enums.Enums.Media;
using WebApi.Enums.Enums.Sponsor;

namespace WebApi.Models.Models.Sponsor;

public class Advertisement
{
    [JsonPropertyName("id")]
    public long Id { get; set; }

    [JsonPropertyName("sponsorId")]
    public long UserId { get; set; }

    [JsonPropertyName("text")]
    public string? Text { get; set; }

    [JsonPropertyName("name")]
    public string? Name { get; set; }

    [JsonPropertyName("targetAudience")]
    public string? TargetAudience { get; set; }

    [JsonPropertyName("media")]
    public string? Media { get; set; }

    [JsonPropertyName("show")]
    public bool Show { get; set; }

    [JsonPropertyName("updated")]
    public bool Updated { get; set; }

    [JsonPropertyName("priority")]
    public AdvertisementPriority Priority { get; set; }

    [JsonPropertyName("mediaType")]
    public MediaType MediaType { get; set; }

    [JsonPropertyName("status")]
	public AdvertisementStatus Status { get; set; }

	[JsonPropertyName("adminId")]
    public long? AdminId { get; set; }

    [JsonPropertyName("tags")]
    public string[]? Tags { get; set; }
}
